spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java

```diff
@@ -479,11 +498,14 @@ protected void handleMessageInternal(Message<?> message) throws Exception {
 				}
 			}
 			else {
-				discardMessage(message);
+				discardMessage(message, lock);
+				noOutput = false;
 			}
 		}
 		finally {
-			lock.unlock();
+			if (noOutput || !this.releaseLockBeforeSend) {
```

Comment:
I believe this `noOutput` is for the case of error.
How about if we `releaseLockBeforeSend` and unlock in the `completeGroup()` before send, but then send fail?
So, `noOutput` remains `true`, but lock is unlocked already anyway.

Isn't going to be:
```
* @throws IllegalMonitorStateException if the current thread does not
*         hold this lock
```
?

Reply:
It's not for the case of error; it's for when we only added a message to a group and didn't produce any output. But, you are right, if sendOutputs fails, we'll try to unlock again.

`RentrantLock` has `isHeldByCurrentThread()` but we can't be assured it will be one of those.

I think the only alternative is a `ThreadLocal`.

Reply:
OK. How about to move `noOutput = false;` to the `final` block:
```
try {
    discardMessage(message, lock);
}
finally {
    noOutput = false;
}

```

Reply:
I think my latest commit solves it.

Reply:
👍 

Let me pull it locally for the final review!